lib/mongo/socket.rb

```diff
@@ -123,6 +123,7 @@ def initialize(family)
     def read(length)
       handle_errors do
         data = read_from_socket(length)
+        raise IOError unless data.length == length
```

Comment:
This should raise the `IOError` if `data.length == 0`. There is the possibility that some data is read but not all of it, which is the reason for the loop below. If that code path isn't ever executed, which is the case with the current check, then we have the chance to be reading partial messages.